### PR TITLE
ci: rc, release, tagged rc, hotfix workflows

### PR DESCRIPTION
**Pull request by awsluja**
#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by sobolk — CHANGES_REQUESTED** [review]
it seems that `publish-codebuild.sh` does not reflect changes made on CCI to split the deployment step into pieces.

This PR should either incorporate these changes OR `release` should be descoped from here and added later with the split.

File: .circleci/config.base.yml [hunk 1]
```diff
@@ -1,1560 +1,1560 @@
-version: 2.1
-orbs:
-  aws-ecr: circleci/aws-ecr@6.15.3
-machine:
-  environment:
-    PATH: '${PATH}:${HOME}/${CIRCLE_PROJECT_REPONAME}/node_modules/.bin'
+# version: 2.1
+# orbs:
```
Comment (sobolk) [inline .circleci/config.base.yml:2]: disabling CCI should be separate PR.

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,268 +1,268 @@
-version: 2.1
+# version: 2.1
 
-# this allows you to use CircleCI's dynamic configuration feature
-setup: true
+# # this allows you to use CircleCI's dynamic configuration feature
```
Comment (sobolk) [inline .circleci/config.yml:3]: same here.

File: .circleci/local_publish_helpers_codebuild.sh [hunk 3]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+git checkout $BRANCH_NAME
+
```
Comment (sobolk) [inline .circleci/local_publish_helpers_codebuild.sh:4]: We shouldn't checkout a branch every time we load this script.
Is this intentional?

File: .circleci/local_publish_helpers_codebuild.sh [hunk 4]
```diff
@@ -23,48 +25,37 @@ function startLocalRegistry {
...
     export version=$(./amplify-pkg-linux-x64 --version)
 
-    if [[ "$CIRCLE_BRANCH" == "release" ]] || [[ "$CIRCLE_BRANCH" =~ ^run-e2e-with-rc\/.* ]] || [[ "$CIRCLE_BRANCH" =~ ^release_rc\/.* ]] || [[ "$CIRCLE_BRANCH" =~ ^tagged-release ]]; then
-        aws --profile=s3-uploader s3 cp amplify-pkg-win-x64.tgz s3://aws-amplify-cli-do-not-delete/$(echo $version)/amplify-pkg-win-x64-$(echo $hash).tgz
-        aws --profile=s3-uploader s3 cp amplify-pkg-macos-x64.tgz s3://aws-amplify-cli-do-not-delete/$(echo $version)/amplify-pkg-macos-x64-$(echo $hash).tgz
-        aws --profile=s3-uploader s3 cp amplify-pkg-linux-arm64.tgz s3://aws-amplify-cli-do-not-delete/$(echo $version)/amplify-pkg-linux-arm64-$(echo $hash).tgz
-        aws --profile=s3-uploader s3 cp amplify-pkg-linux-x64.tgz s3://aws-amplify-cli-do-not-delete/$(echo $version)/amplify-pkg-linux-x64-$(echo $hash).tgz
+    if [[ "$BRANCH_NAME" == "release" ]] || [[ "$BRANCH_NAME" =~ ^run-e2e-with-rc\/.* ]] || [[ "$BRANCH_NAME" =~ ^release_rc\/.* ]] || [[ "$BRANCH_NAME" =~ ^tagged-release ]]; then
+        aws s3 cp amplify-pkg-win-x64.tgz s3://$PKG_CLI_BUCKET_NAME/$(echo $version)/amplify-pkg-win-x64-$(echo $hash).tgz
```
Comment (sobolk) [inline .circleci/local_publish_helpers_codebuild.sh:34]: Can we make a copy of this file instead? and drop non-CB version together with CCI later?

Please do this for all other `.sh` files where we made such assumption.

File: .circleci/publish-step-2-verdaccio.sh [hunk 5]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash -e
 
+source ./.circleci/codebuild-checkout.sh
+
```
Comment (sobolk) [inline .circleci/publish-step-2-verdaccio.sh:4]: This is unexpected.
In CCI `.circleci/publish-step-1-set-versions.sh` adds a commit to local repo that is NOT pushed to git.
That commit is retained in other jobs because we cache result of `build` step, including `.git` directory which preserves that information. We should not need to checkout everywhere if we we cache workspace properly.

This is concern for other files that attempt to checkout after `build` step.

File: codebuild_specs/release_workflows/release_workflow.yml [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        - verify_pkg_cli
+    - identifier: post_publish_push_to_git
+      buildspec: codebuild_specs/release_workflows/post_publish_push_to_git.yml
+      depend-on:
+        - publish_to_npm
+    - identifier: github_release
+      buildspec: codebuild_specs/release_workflows/github_release.yml
+      env:
+        compute-type: BUILD_GENERAL1_LARGE
+      debug-session: true
```
Comment (sobolk) [inline codebuild_specs/release_workflows/release_workflow.yml:90]: do we want these debug sessions ?

**Review by sobolk — COMMENTED** [review]
Did get a chance to test it? We should at least get successful tagged release using this workflow.

Major concerns:
- We should decouple CB release workflows from CCI sunset. These should be separate changes. Ability to use both systems should exists until we gain confidence in CB. In event of emergency we should be able to buy CCI credits and revert back to old systems should CB prevent us from shipping. Smooth migration must include period where both old and new system is accessible.
- The necessity of `codebuild-checkout` in release steps 2,3,4. Revised release process assumes that outcome of `build` step is cached **including** git metadata. Can we get this fixed so that checkout is necessary only in entry steps (build, build-tests-standalone? )

File: .circleci/config.base.yml (hunk 1 above)
Comment (edwardfoyle) [reply .circleci/config.base.yml:2]: Yep, we shouldn't turn off CCI until we've done at least a couple releases using CodeBuild

File: .circleci/local_publish_helpers_codebuild.sh (hunk 4 above)
Comment (awsluja) [reply .circleci/local_publish_helpers_codebuild.sh:32]: done

File: codebuild_specs/amplify_general_config_tests.yml [hunk 7]
```diff
@@ -10,3 +10,6 @@ phases:
   build:
     commands:
       - source ./shared-scripts.sh && _amplifyGeneralConfigTests
+artifacts:
+  files:
+    - 'shared-scripts.sh'
```
Comment (sobolk) [inline codebuild_specs/amplify_general_config_tests.yml:15]: what is this for ?

File: codebuild_specs/publish_to_npm.yml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+version: 0.2
+env:
+  shell: bash
+  git-credential-helper: yes
```
Comment (sobolk) [inline codebuild_specs/publish_to_npm.yml:4]: how does`git-credential-helper` work ?

**Review by sobolk — APPROVED** [review]
LGTM.

File: .circleci/cb-publish-step-3-npm.sh [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+# this causes the script to keep running even after failed publishes
+# this function forces failed publishes to exit on failure
+function lernaPublishExitOnFailure {
+  # exit on failure
+  set -e
+  # run lerna publish with the args that were passed to this function
+  # duplicate stdout to a temp file
+  # grep the temp file for the lerna err token and return exit 1 if found (-v option inverts grep exit code)
+  npx lerna publish "$@" | tee /tmp/publish-results && grep -qvz "lerna ERR!" < /tmp/publish-results
+}
```
Comment (jhockett) [inline .circleci/cb-publish-step-3-npm.sh:13]: Can this function live in `.circleci/local_publish_helpers_codebuild.sh` instead of duplicating it?

File: shared-scripts.sh [hunk 10]
```diff
@@ -165,17 +166,16 @@ function _publishToLocalRegistry {
...
+    ./.circleci/cb-publish-step-2-verdaccio.sh
     unsetNpmRegistryUrl
 
     echo Generate Change Log
     # Leaving this breadcrumb here "git reset --soft HEAD~1"
-    # we commented this out because the publish script is now checking out the current branch, and this started to fail as a result
+    # we commented this out previously because the publish script is now checking out the current branch, and this started to fail as a result
     # if we run into problems in the future, we should revisit this
-    # git reset --soft HEAD~1
+    git reset --soft HEAD~1
```
Comment (jhockett) [inline shared-scripts.sh:178]: What's the goal of doing a soft reset?

File: .circleci/cb-publish-step-3-npm.sh (hunk 9 above)
Comment (awsluja) [reply .circleci/cb-publish-step-3-npm.sh:13]: yes we can refactor after; I actually duplicated several cb-<name> files per Kamil's request to not touch existing .sh files

File: codebuild_specs/amplify_general_config_tests.yml (hunk 7 above)
Comment (awsluja) [reply codebuild_specs/amplify_general_config_tests.yml:15]: codebuild requires a job to output an artifact, otherwise other jobs cannot depend on it downstream

File: codebuild_specs/publish_to_npm.yml (hunk 8 above)
Comment (awsluja) [reply codebuild_specs/publish_to_npm.yml:4]: only the jobs that require .git credentials should use git-credential-helper; this injects the GH credentials via the git-credential-helper https://git-scm.com/docs/gitcredentials so that things like push work. Can only be used on jobs that are manually triggered via console/api.

File: codebuild_specs/release_workflows/release_workflow.yml (hunk 6 above)
Comment (awsluja) [reply codebuild_specs/release_workflows/release_workflow.yml:90]: this is just part of it; the other part is to manually trigger the job with debug sessions enabled & ssm enabled

File: shared-scripts.sh (hunk 10 above)
Comment (awsluja) [reply shared-scripts.sh:178]: if i remember correctly; its to remove the commit that is created during publish to verdaccio step so that it doesn't affect the unified changelog step

Comment (sobolk) [reply shared-scripts.sh:178]: this shouldn't be necessary. does CCI do this ?

Comment (sobolk) [reply shared-scripts.sh:178]: or maybe it does make some sense, should we unreset after step which is affected by this ?

Comment (jhockett) [reply shared-scripts.sh:178]: I see, this unified changelog script is designed to work with staged files:

```ts

 // This script is intended to be run after running `yarn publish-to-verdaccio` when all CHANGELOG file changes are staged but not committed
 // It will scan all staged CHANGELOG.md files and merge them into a single UNIFIED_CHANGELOG file
 ```

**Review by jhockett — APPROVED** [review]

File: codebuild_specs/amplify_general_config_tests.yml (hunk 7 above)
Comment (sobolk) [reply codebuild_specs/amplify_general_config_tests.yml:15]: :-)
